app/scripts/offscreen.js

```diff
@@ -5,6 +5,30 @@ import {
 } from '../../shared/constants/offscreen-communication';
 import { getSocketBackgroundToMocha } from '../../test/e2e/background-socket/socket-background-to-mocha';
 
+/**
+ * Returns whether the offscreen document already exists or not.
+ *
+ * See https://developer.chrome.com/docs/extensions/reference/api/offscreen#before_chrome_116_check_if_an_offscreen_document_is_open
+ *
+ * @returns True if the offscreen document already is has been opened, otherwise false.
+ */
+async function hasOffscreenDocument() {
+  const { chrome, clients } = globalThis;
+  // getContexts is only available in Chrome 116+
+  if ('getContexts' in chrome.runtime) {
+    const contexts = await chrome.runtime.getContexts({
+      contextTypes: ['OFFSCREEN_DOCUMENT'],
+    });
+    return contexts.length > 0;
+  }
+  const matchedClients = await clients.matchAll();
+  return matchedClients.some(
+    (client) =>
+      client.url.includes(chrome.runtime.id) &&
+      client.url.endsWith('offscreen.html'),
```

Comment:
Might not be relevant here, but if any of the `client` here uses an URL with GET parameters like:

```
https://foobar.com?fake=offscreen.html
```

This might also return `true` in this case.

I have 0 context in this code, so this might not be relevant/possible. 😅

If that's a potential problem, then we could probably use `URL.origin` instead?

Reply:
As far as I can tell it only returns clients "owned" by the service worker. So it'll strictly be windows owned by the extension, which just seems to be `chrome-extension://id/something.html`. I modified the example from the Google docs slightly for more safety.

Regardless, 99.9% of users will use the if-clause assuming they aren't on an ancient browser.

Reply:
Simplified this a bit now, should be more safe like this.

Reply:
Thanks, looks great!